delivery_purchase_label/models/purchase_order.py

```diff
@@ -0,0 +1,93 @@
+        "stock.picking",
+        "Delivery Label Picking",
+        store=True,
+        readonly=True,
+    )
+
+    @api.model
+    def _states_to_generate_delivery_label(self):
+        """Labels will be (re)generated only if the PO is in one of these states."""
+        return ["draft", "sent"]
+
+    @api.model
+    def _mail_templates_to_not_attach_labels(self):
+        return self.env.ref("purchase.email_template_edi_purchase_reminder").ids
+
+    def action_rfq_send(self):
+        self.ensure_one()
+        self._generate_purchase_delivery_label()
+        return super().action_rfq_send()
+
+    def _is_valid_for_vendor_labels(self):
+        self.ensure_one()
+        if self.state not in self._states_to_generate_delivery_label():
+            return
+        if not self.dest_address_id:
+            return False
+        if not any(
+            product.type in ["product", "consu"]
+            for product in self.order_line.product_id
+        ):
+            return False
+        return True
+
+    def _generate_purchase_delivery_label(self):
+        """Create a transfer to generate the carrier labels."""
+        self.ensure_one()
+        if not self._is_valid_for_vendor_labels():
+            return
+        # Find the carrier that will be used
+        carrier = self.partner_id.purchase_delivery_carrier_id
+        if not carrier.purchase_label_picking_type:
+            return
+        if not self.partner_id.property_stock_supplier.id:
+            raise UserError(
+                _(
+                    "You must set a Vendor Location for this partner %s",
+                    self.partner_id.name,
+                )
+            )
+        order = self.with_company(self.company_id)
+        # Create and process the transer to send the labels
+        values = order._get_purchase_delivery_label_picking_value(carrier)
+        picking = self.env["stock.picking"].with_user(SUPERUSER_ID).create(values)
+        moves = order.order_line._create_stock_moves(picking)
+        # Remove the link on the sale and purchase
+        # To not impact the delivered quantity on them
+        picking.sale_id = False
+        moves.sale_line_id = False
+        moves.purchase_line_id = False
+        picking.action_assign()
+        for move in moves:
+            move.quantity_done = move.product_uom_qty
+        picking._action_done()
+        order.delivery_label_picking_id = picking
+        picking.message_post_with_view(
+            "mail.message_origin_link",
+            values={"self": picking, "origin": self},
+            subtype_id=self.env.ref("mail.mt_note").id,
+        )
+
+    def _get_purchase_delivery_label_picking_value(self, carrier):
+        return {
+            "picking_type_id": carrier.purchase_label_picking_type.id,
+            "partner_id": self.dest_address_id.id,
+            "user_id": False,
+            "date": fields.Datetime.now(),
+            "origin": self.name,
```

Comment:
or maybe the SO number ? To clarify
```suggestion
            "origin": self.origin,
```

Reply:
Yes use self.origin this makes it much easier to generate the label out the purchae delivery label record. 
It represents then more or less a copy of the initial purchase order.

Reply:
Don't we wanted both on label? If we concatenate the two doesn't it solve the label?
"%origin - %name" ?

Reply:
ping @TDu